agent/wsclient/client.go

```diff
@@ -198,35 +207,56 @@ func (cs *ClientServerImpl) Connect() error {
 			}
 		}
 		seelog.Warnf("Error creating a websocket client: %v", err)
-		return fmt.Errorf(string(resp) + ", " + err.Error())
+		return errors.Wrapf(err, "websocket client: unable to dial %s response: %s",
+			parsedURL.Host, string(resp))
 	}
+
+	cs.writeLock.Lock()
+	defer cs.writeLock.Unlock()
+
 	cs.conn = websocketConn
+	seelog.Debugf("Established a Websocket connection to %s", cs.URL)
 	return nil
 }
 
 // IsReady gives a boolean response that informs the caller if the websocket
 // connection is fully established.
 func (cs *ClientServerImpl) IsReady() bool {
-	cs.writeLock.Lock()
-	defer cs.writeLock.Unlock()
+	cs.writeLock.RLock()
+	defer cs.writeLock.RUnlock()
+
 	return cs.conn != nil
 }
 
-// SetConnection passes a websocket connection object into the client.
+// SetConnection passes a websocket connection object into the client. This is used only in
+// testing and should be avoided in non-test code.
 func (cs *ClientServerImpl) SetConnection(conn WebsocketConn) {
 	cs.conn = conn
 }
 
+// SetReadDeadline sets the read deadline for the websocket connection
+// A read timeout results in an io error if there are any outstanding reads
+// that exceed the deadline
+func (cs *ClientServerImpl) SetReadDeadline(t time.Time) error {
```

Comment:
Would it be better to pass the duration and calculate the time right before calling the `SetReadDeadline`, it's more accurate, as the passed in time could already beyond time.Now() for some wired reason.